[clang] Add fixed point precision macros

PiJoules opened the pull request:
This defines the builtin macros specified in `7.18a.3 Precision macros` of ISO/IEC TR 18037:2008. These are the `__*__` versions of them and the formal definitions in stdfix.h can use them.

PiJoules committed:
[clang] Add fixed point precision macros

This defines the builtin macros specified in `7.18a.3 Precision macros`
of ISO/IEC TR 18037:2008. These are the `__*__` versions of them and the formal
definitions in stdfix.h can use them.

PiJoules commented:
cc @lntue

PiJoules commented:
ping

AaronBallman commented on clang/test/Preprocessor/fixed-point.c:
```diff
@@ -0,0 +1,67 @@
+// CHECK-DAG:#define __SFRACT_MAX__ 0.9921875hr
+// CHECK-DAG:#define __SFRACT_MIN__ (-0.5hr-0.5hr)
+
+// CHECK-DAG:#define __USFRACT_EPSILON__ 0.00390625uhr
+// CHECK-DAG:#define __USFRACT_FBIT__ 8
+// CHECK-DAG:#define __USFRACT_MAX__ 0.99609375uhr
```
Values from TR 18037:2008 are:
```
USFRACT_FBIT 7
USFRACT_MAX 0.9921875UHR
USFRACT_EPSILON 0.0078125UHR
```
The values are conforming though different, just checking whether this is intentional given that `SFRACT` are using the identical values from the TR (this same thing happens with all the other `unsigned` variants).

PiJoules replied:
I believe those values are just example values for `7.18a.3 Precision macros` and the actual values are implementation defined:

```
The values in the following list shall be replaced by constant expressions with implementation-defined values with the same type
```

The values in this test come from the "typical desktop processor" example in `A.3 Possible Data Type Implementations`.

AaronBallman replied:
I understood that they were implementation-defined, but I had missed we were pulling the values from `A.3`. Thanks for the explanation!

AaronBallman commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -1097,6 +1150,47 @@ static void InitializePredefinedMacros(const TargetInfo &TI,
+                           TI.getFractScale(), /*Signed=*/true);
+    DefineFixedPointMacros(TI, Builder, "UFRACT", "ur", TI.getFractWidth(),
+                           TI.getUnsignedFractScale(), /*Signed=*/false);
+    DefineFixedPointMacros(TI, Builder, "LFRACT", "lr", TI.getLongFractWidth(),
```
Slight preference for using `L` instead of `l` because of font problems where the `l` may look like a `1`. Might be worth making them use the uppercase suffix consistently.

PiJoules replied:
Done

AaronBallman reviewed (commented):
I'm by no means a fixed-point expert, so this is a bit of a drive-by review.

One thing to note is that we shouldn't be using N1169 because TR 18037 is freely available: https://standards.iso.org/ittf/PubliclyAvailableStandards/c051126_ISO_IEC_TR_18037_2008.zip

The changes should come with a release note so that users know about them.

PiJoules commented:
> I'm by no means a fixed-point expert, so this is a bit of a drive-by review.
> 
> One thing to note is that we shouldn't be using N1169 because TR 18037 is freely available: https://standards.iso.org/ittf/PubliclyAvailableStandards/c051126_ISO_IEC_TR_18037_2008.zip
> 
> The changes should come with a release note so that users know about them.

Replaced mentions of N1169 with TR 18037. Also updated clang release notes.

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -297,6 +297,12 @@ DWARF Support in Clang
 Floating Point Support in Clang
 -------------------------------
 
+Fixed Point Support in Clang
+-------------------------------
```
```suggestion
Fixed Point Support in Clang
-----------------------------
```

AaronBallman reviewed (approved):
Aside from a minor nit with the documentation, LGTM!

MaskRay commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -768,6 +768,59 @@ void InitializeOpenCLFeatureTestMacros(const TargetInfo &TI,
+    std::string Literal;
+    std::string HalfStr = ConstructFixedPointLiteral(Val.shr(1), Suffix);
+    Literal.push_back('(');
+    Literal += HalfStr;
```
Returning `SmallString`  and using `Twine` to concatenate strings is slightly more efficient.
